source/common/stats/statsd.cc

```diff
@@ -97,12 +97,55 @@ TcpStatsdSink::TlsSink::~TlsSink() {
   }
 }
 
+void TcpStatsdSink::TlsSink::beginFlush(bool expect_empty_buffer) {
+  ASSERT(!expect_empty_buffer || buffer_.length() == 0);
+  ASSERT(current_slice_mem_ == nullptr);
+  UNREFERENCED_PARAMETER(expect_empty_buffer);
+
+  uint64_t num_iovecs = buffer_.reserve(FlushSliceSizeBytes, &current_buffer_slice_, 1);
+  ASSERT(num_iovecs == 1);
+  UNREFERENCED_PARAMETER(num_iovecs);
+
+  ASSERT(current_buffer_slice_.len_ >= FlushSliceSizeBytes);
+  current_slice_mem_ = reinterpret_cast<char*>(current_buffer_slice_.mem_);
+}
+
+void TcpStatsdSink::TlsSink::commonFlush(const std::string& name, uint64_t value, char stat_type) {
+  ASSERT(current_slice_mem_ != nullptr);
+  // 40 > 6 (prefix) + 4 (random chars) + 30 for number (bigger than it will ever be)
+  if (current_buffer_slice_.len_ - usedBuffer() < name.size() + 40) {
+    endFlush(false);
+    beginFlush(false);
+  }
+
+  // Produces something like "envoy.{}:{}|c\n"
+  memcpy(current_slice_mem_, StatPrefix, sizeof(StatPrefix) - 1);
+  current_slice_mem_ += sizeof(StatPrefix) - 1;
+  memcpy(current_slice_mem_, name.c_str(), name.size());
+  current_slice_mem_ += name.size();
+  *current_slice_mem_++ = ':';
+  current_slice_mem_ += StringUtil::itoa(current_slice_mem_, 30, value);
+  *current_slice_mem_++ = '|';
+  *current_slice_mem_++ = stat_type;
```

Comment:
I assume we're doing C-style memory fill here for performance reasons (otherwise I'd be in favor of some kind of string buffer and then write).